<a href="https://colab.research.google.com/github/fernandolievano/procesamiento-del-habla/blob/main/%5BPH%5D_TP3_An%C3%A1lisis_similitud_de_sem%C3%A1ntica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Trabajo Práctico 3 - Análisis de similitud de semántica
### Procesamiento del Habla
---
## Objetivo:
 Explorar y comprender las relaciones semánticas entre palabras mediante el cálculo de la similitud del coseno utilizando modelos de lenguaje.

### Setup del entorno

In [ ]:
# instalación de librerías
!pip install spacy

In [ ]:
# descarga del modelo
!python -m spacy download es_core_news_md

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 20.0 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('es_core_news_md')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [ ]:
# import y configuraciones iniciales
import spacy

# ccarga de modelo en español con vectores preentrenados
nlp = spacy.load("es_core_news_md")

In [ ]:
# funciónes reutilizables

def better_print(message: str):
    print("===" * 30)
    print(f' {message}')
    print("===" * 30)

### ¿Que significa Embeddings?

Los **embeddings** son representaciones numéricas de palabras o textos que permiten a una computardora interpretar su significado. Cada palabra se transforma en un vector matemático donde palabras con significados o contextos similares quedan más cerca entre sí.

Esto ayudo a los modelos de lenguaje a comprender relaciones semánticas y relizar tareas como medir similitud entre palabras, clasificación de texto o traducción automática.

*(1)*

### Similitud del coseno

La *similitud del coseno* es una medida matemática utilizada en Natural Language Processing para comparar qué tan parecidos son dos vectores, como los **embeddings de palabras o textos**. Funciona calculando el ángulo entre ambos vectores: mientras más pequeño sea el ángulo, mayor será la similitud semántica entre ellos.

Sus valores suelen ir entre -1 y 1. Un valor cercano a 1 indica que los vectores son muy similares, 0 indica que no tienen relación significativa y un valor cercano a -1 indica que son opuestos. En NLP normalmente los valores más altos significan que las palabras o textos tienen significados o contextos parecidos.

*(2)*

## Similitud entre palabras

### Similitud Positiva

Programador - Desarrollador

In [ ]:
palabra1 = nlp('Programador')
palabra2 = nlp('Desarrollador')

similaridad = palabra1.similarity(palabra2)

better_print(f'La similitud entre {palabra1} y {palabra2} es: {similaridad}')

 La similitud entre Programador y Desarrollador es: 0.7910242676734924


 Las palabras **programador** y **desarrollador** obtuvieron una similitud de 0.791 lo que representa una similitud semántica alta (cercana a 1).
 Esto pasa porque ambos términos se suelen relacionar a la informática y suelen utilizarse en contextos similares, como el desarrollo de software.

### Similitud Neutra

Ratón - Joystick

In [ ]:
palabra3 = nlp('Ratón')
palabra4 = nlp('Joystick')

similaridad = palabra3.similarity(palabra4)

better_print(f'La similitud entre {palabra3} y {palabra4} es: {similaridad}')

 La similitud entre Ratón y Joystick es: -0.04088244214653969


Las palabras **ratón** y **joystick** obtuvieron una similitud de -0.04, es decir, un valor muy cercano a 0.

Esto indica una relación semántica neutra entre ambos términos. Aunque ambos pertenecen al ámbito de los periféricos de PC, suelen utilizarse en contextos diferentes y no comparten un significado directamente relacionado.

### Similitud Negativa

Vector - Stream

In [101]:
# palabra5 = nlp('Tanque')
palabra5 = nlp('Antibalas')
# palabra6 = nlp('Adhoc')
palabra6 = nlp('Adhoc')

similaridad = palabra5.similarity(palabra6)

better_print(f'La similitud entre {palabra5} y {palabra6} es: {similaridad}')

 La similitud entre Antibalas y Adhoc es: -0.16573049128055573


La similitud del coseno entre **Adhoc** y **Antibalas** fue de -0.16, lo que indica una relación semántica muy baja entre ambos términos. Lo que sugiere que las palabras suelen aparecer en contextos diferentes, o sea, no suelen compartir un significado o asociación relevante.

## Análisis de Similitud Semántica por Pares de Palabras

### 1. Programador - Desarrollador
- **Similitud:** 0.791
- **Análisis:** Este par de palabras obtuvo una similitud muy alta (cercana a 1). Esto se debe a que `Programador` y `Desarrollador` son términos casi sinónimos en el ámbito de la informática y el desarrollo de software. Ambos roles implican la creación y mantenimiento de código, y a menudo se usan indistintamente o con pequeñas diferencias de especialización que no los alejan semánticamente en un contexto general. El modelo de lenguaje las identifica como fuertemente relacionadas debido a su co-ocurrencia frecuente en contextos similares.

### 2. Ratón - Joystick
- **Similitud:** -0.040
- **Análisis:** La similitud entre `Ratón` y `Joystick` es muy cercana a 0, lo que indica una relación semántica neutra. Aunque ambos son periféricos de entrada utilizados para interactuar con computadoras o videojuegos, sus funciones principales y los contextos de uso son bastante distintos. Un ratón se asocia más con la navegación general y el trabajo de oficina, mientras que un joystick está fuertemente ligado a los videojuegos. A pesar de su categoría común (periféricos), el modelo no encuentra una superposición significativa en sus contextos semánticos específicos.

### 3. Antibalas - Adhoc
- **Similitud:** -0.165
- **Análisis:** Este par muestra una similitud baja y negativa, lo que sugiere una relación semántica débil o incluso opuesta en ciertos contextos. `Antibalas` se refiere a algo diseñado para resistir proyectiles, implicando protección, seguridad y materiales resistentes. `Adhoc`, por otro lado, significa "para un propósito particular o específico" y no tiene una connotación material o de seguridad intrínseca, sino más bien de improvisación o solución temporal. Los contextos en los que aparecen estas palabras son generalmente muy diferentes, y el modelo de lenguaje no encuentra conexiones relevantes entre ellas, resultando en una similitud cercana a la no relación o disimilitud.

## Referencias

### Fuentes

1. https://www.ibm.com/think/topics/word-embeddings
2. https://www.ibm.com/think/topics/cosine-similarity

### Uso de IA

Se utilizó IA para generar el scaffolding para que el notebook tenga una estructura clara y limpia.

[Conversación ChatGPT](https://chatgpt.com/share/6a022f70-e358-83e9-834d-56f59befaa11).